core/src/main/java/com/yahoo/imapnio/client/ImapClientRespReader.java

```diff
@@ -44,79 +56,81 @@ public ImapClientRespReader(final int maxLineLength) {
      * and then continue read the next CRLF.
      */
     @Override
-    protected Object decode(final ChannelHandlerContext ctx, final ByteBuf buffer) throws Exception {
-        final int bufferBytes = buffer.readableBytes();
-        if (bufferBytes > 0) {
-            ByteBuf lineBuffer = null;
-            do {
-                lineBuffer = (ByteBuf) super.decode(ctx, buffer);
-                // lineBuffer might be null(no delimiter in the buffer) when receiving partial response,
-                // skip current decoding and wait until next CRLF.
-                if (lineBuffer != null) {
-                    final int readableBytes = lineBuffer.readableBytes();
-
-                    if (literalCount > 0) {
-                        // Continue buffering literal response.
-                        literalCount = literalCount - readableBytes;
-                        if (literalCount <= 0) {
-                            // Already read all literal response, check if current line contains new literal response.
-                            literalCount = getLiteralCount(lineBuffer, readableBytes, readableBytes + literalCount);
-                        }
-                        literalBuffer.writeBytes(lineBuffer);
-                    } else {
-                        literalCount = getLiteralCount(lineBuffer, readableBytes, 0);
-                        if (literalCount > 0) {
-                            // Init literal buffer when receiving first literal response.
-                            literalBuffer = Unpooled.buffer(readableBytes + literalCount);
-                            literalBuffer.writeBytes(lineBuffer);
-                        }
+    protected Object decode(final ChannelHandlerContext ctx, final ByteBuf inputBuf) throws Exception {
+        if (inputBuf.readableBytes() <= 0) {
+            return null;
+        }
+        if (literalBuf == null) { // initialized small buffer first, ByteBuf itself handles the growing
+            literalBuf = Unpooled.buffer(INITIAL_BUF_LEN);
```

Comment:
With literal you already know the size, why can't we pre-allocate when detected instead of additional logic?

Reply:
done